Data/Primitive/ByteArray.hs

```diff
@@ -126,22 +129,49 @@ newAlignedPinnedByteArray (I# n#) (I# k#)
 -- | Yield a pointer to the array's data. This operation is only safe on
 -- /pinned/ byte arrays allocated by 'newPinnedByteArray' or
 -- 'newAlignedPinnedByteArray'.
+--
+-- Prefer 'withByteArrayContents', which ensures that the array is not
+-- garbage collected while the pointer is being used.
 byteArrayContents :: ByteArray -> Ptr Word8
 {-# INLINE byteArrayContents #-}
 byteArrayContents (ByteArray arr#) = Ptr (byteArrayContents# arr#)
 
+-- | A composition of 'byteArrayContents' and 'keepAliveUnlifted'.
+-- The callback function must not return the pointer.
+withByteArrayContents :: PrimBase m => ByteArray -> (Ptr Word8 -> m a) -> m a
+{-# INLINE withByteArrayContents #-}
+withByteArrayContents (ByteArray arr#) f =
+  keepAliveUnlifted arr# (f (Ptr (byteArrayContents# arr#)))
```

Comment:
I used to think the lifted `keepAlive` would be unsafe here, but as I've thought about it more, I have realized that it should be safe since it would cause the `ByteArray#` inside the `ByteArray` to transitively be kept alive. I would say that the choice to use `keepAliveUnlifted` is made for performance, not correctness. If we use `keepAlive`, it's forces the wrapper data constructor `ByteArray` to actually be materialized on the heap. GHC's optimizations play more nicely with `keepAliveUnlifted`. I do need to document this though.

Reply:
I've documented this.